core/container.go

```diff
@@ -286,26 +286,64 @@ func (container *Container) Mounts(ctx context.Context) ([]string, error) {
 }
 
 func (container *Container) Directory(ctx context.Context, gw bkgw.Client, dirPath string) (*Directory, error) {
+	dir, err := locatePath(ctx, container, dirPath, gw, NewDirectory)
+	if err != nil {
+		return nil, err
+	}
+
+	// check that the directory actually exists so the user gets an error earlier
+	// rather than when the dir is used
+	_, err = dir.Stat(ctx, gw, ".")
+	if err != nil {
+		return nil, err
+	}
+
+	return dir, nil
+}
+
+func (container *Container) File(ctx context.Context, gw bkgw.Client, dirPath string) (*File, error) {
+	file, err := locatePath(ctx, container, dirPath, gw, NewFile)
+	if err != nil {
+		return nil, err
+	}
+
+	// check that the file actually exists so the user gets an error earlier
+	// rather than when the file is used
+	_, err = file.Stat(ctx, gw)
+	if err != nil {
+		return nil, err
+	}
+
+	return file, nil
+}
+
+func locatePath[T *File | *Directory](
+	ctx context.Context,
+	container *Container,
+	containerPath string,
+	gw bkgw.Client,
+	init func(context.Context, llb.State, string, specs.Platform) (T, error),
+) (T, error) {
 	payload, err := container.ID.decode()
 	if err != nil {
 		return nil, err
 	}
 
-	dirPath = absPath(payload.Config.WorkingDir, dirPath)
+	containerPath = absPath(payload.Config.WorkingDir, containerPath)
 
-	var dir *Directory
+	var found T
 
 	// NB(vito): iterate in reverse order so we'll find deeper mounts first
 	for i := len(payload.Mounts) - 1; i >= 0; i-- {
 		mnt := payload.Mounts[i]
 
-		if dirPath == mnt.Target || strings.HasPrefix(dirPath, mnt.Target+"/") {
+		if containerPath == mnt.Target || strings.HasPrefix(containerPath, mnt.Target+"/") {
```

Comment:
Once this code is combined w/ the secret PR where we can mount secrets as files at paths, does that mean that it would be possible to ask for a file at the path where the secret is mounted but then get a different file here? Or do secret file mounts never override existing files?

I would need to think a bit more about what the correct behavior here should be, just a possible corner case to be aware of for now (or maybe not an issue at all).

Reply:
Hm true, I was on the fence as to whether whether secret files should be mounts or a separate field, since they're represented differently in LLB and can optionally be env too, but this feels like a solid argument for having secret files be mounts. And then we'd just need another field for secret env which isn't a huge deal. I'll look into this now.

Reply:
Ok, I take that back, the code feels a lot more natural with them separate since they really don't behave like mounts in LLB. Even if I put them in the list of mounts in the container ID payload I'll have to loop over all the mounts first to find the secrets for their `RunOptions`, call `.Run`, and then process the mounts (_ignoring_ any secrets) against the state returned by `Run`, so it's no different from just having them separate in the first place.

Gonna write a test for this and see what happens against a branch with the two merged. I expect the command itself to see the secret mount, and if another mount stacks on top of it `file` will return that, and if not I'm guessing it might return the empty base mount of the secret from the rootfs. If it finds anything at all.

Reply:
Added a test for this.

With a file mount + secret, `container.file` finds the file, not the secret. Without a file mount and only a secret, `container.file` finds nothing and you get a 'file not found' error.

I'm not sure if I'd declare that "correct" but the test should at least cover the current behavior in case people rely on it and it suddenly changes somehow.

Reply:
> I'm not sure if I'd declare that "correct" but the test should at least cover the current behavior in case people rely on it and it suddenly changes somehow.

Sounds great, agree I don't know if it's "correct" but I think that behavior is probably slightly safer to default to in that you can't just accidentally read a secret into plaintext. So correct enough I guess.